Add option to enable/disable a single DRR

NicerNewerCar opened the pull request:
* Closes #206 
* Adds a toggle to each volume to allow the user to switch on and off which volumes are being displayed.
    * Note: If a volume is toggled off, it will be disabled visually and during any optimizations. 
![toggle-drrs](https://github.com/BrownBiomechanics/Autoscoper/assets/30351234/48e75a52-f9e9-492d-9c3a-08008bd877c3)

- [ ] Fix this UI bug
![image](https://github.com/BrownBiomechanics/Autoscoper/assets/30351234/ecdb08b6-fce0-4833-a7c3-249efd681918)

bpaniagua reviewed (approved):
lgtm

jcfr commented on autoscoper/src/ui/AutoscoperMainWindow.cpp:
```diff
@@ -594,7 +594,7 @@ void AutoscoperMainWindow::setupUI()
   for (unsigned int i = 0; i < tracker->trial()->volumes.size(); i++) {
     std::cout << "Volume Name: " << tracker->trial()->volumes[i].name() << std::endl;
-    volumes_widget->addVolume(tracker->trial()->volumes[i].name());
+    volumes_widget->addVolume(tracker->trial()->volumes[i].name(),i);
```
```suggestion
    volumes_widget->addVolume(tracker->trial()->volumes[i].name(), i);
```

jcfr commented on autoscoper/src/ui/VolumeDockWidget.cpp:
```diff
@@ -42,12 +44,15 @@ void VolumeDockWidget::clearVol(){
+    renderers.push_back(view->drrRenderer(idx));
+  }
 
+  QListWidgetItem* volumeItem = new VolumeListWidgetItem(dock->listWidget,fi.completeBaseName(),mainwindow, &renderers);
```
```suggestion
  QListWidgetItem* volumeItem = new VolumeListWidgetItem(dock->listWidget, fi.completeBaseName(), mainwindow, &renderers);
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,52 @@
+#include "ui/AutoscoperMainWindow.h"
+
+#include <QCheckBox>
+#include <QLabel>
+#include <QDockWidget>
```
```suggestion
#include <QDockWidget>
#include <QLabel>
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,52 @@
+  setup(listWidget);
+}
+
+
```
```suggestion
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,52 @@
+  visibleCheckBox_->connect(visibleCheckBox_, SIGNAL(stateChanged(int)), this, SLOT(on_visibleCheckBox__stateChanged(int)));
+}
+
+
```
```suggestion
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.h:
```diff
@@ -0,0 +1,31 @@
+#include <QObject>
+#include <QListWidgetItem>
+
+class QCheckBox;
+class AutoscoperMainWindow;
```
```suggestion
class AutoscoperMainWindow;
class QCheckBox;
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.h:
```diff
@@ -0,0 +1,31 @@
+#ifndef VOLUMELISTWIDGETITEM_H
+#define VOLUMELISTWIDGETITEM_H
+
+#include <QObject>
+#include <QListWidgetItem>
```
```suggestion
#include <QListWidgetItem>
#include <QObject>
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,52 @@
+
+VolumeListWidgetItem::VolumeListWidgetItem(QListWidget* listWidget,QString name, AutoscoperMainWindow* main_window, std::vector< xromm::gpu::RayCaster*>* renderers) : QListWidgetItem(listWidget) {
+  this->name_ = name;
+  visible_ = true;
```
```suggestion
```

This is redundant with the state already associated with the checkbox and RayCaster

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,52 @@
+
+
+void VolumeListWidgetItem::on_visibleCheckBox__stateChanged(int state) {
+  visible_ = (bool)state;
```
```suggestion
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,52 @@
+  visible_ = (bool)state;
+  if (renderers_.size() > 0) {
+    for (xromm::gpu::RayCaster* renderer : renderers_) {
+      renderer->setVisible(visible_);
```
```suggestion
      bool visible = (bool)state;
      renderer->setVisible(visible);
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.h:
```diff
@@ -0,0 +1,31 @@
+  std::vector< xromm::gpu::RayCaster*> renderers_;
+  QCheckBox *visibleCheckBox_;
+  QString name_;
+  bool visible_;
```
```suggestion
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,48 @@
+  pLayout->setMargin(5);
+  pFrame->setLayout(pLayout);
+  listWidget->setItemWidget(this, pFrame);
+  visibleCheckBox_->connect(visibleCheckBox_, SIGNAL(stateChanged(int)), this, SLOT(on_visibleCheckBox__stateChanged(int)));
```
```suggestion
  QObject::connect(visibilityCheckBox_, SIGNAL(toggled(bool)), this, SLOT(on_visibilityCheckBox__toggled(bool)));
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,48 @@
+  visibleCheckBox_->connect(visibleCheckBox_, SIGNAL(stateChanged(int)), this, SLOT(on_visibleCheckBox__stateChanged(int)));
+}
+
+void VolumeListWidgetItem::on_visibleCheckBox__stateChanged(int state) {
```
```suggestion
void VolumeListWidgetItem::on_visibilityCheckBox__toggled(bool checked) {
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,48 @@
+void VolumeListWidgetItem::on_visibleCheckBox__stateChanged(int state) {
+  if (renderers_.size() > 0) {
+    for (xromm::gpu::RayCaster* renderer : renderers_) {
+      renderer->setVisible((bool)state);
```
```suggestion
      renderer->setVisible(checked);
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.h:
```diff
@@ -0,0 +1,30 @@
+
+  void setup(QListWidget* listWidget);
+public slots:
+  void on_visibleCheckBox__stateChanged(int state);
```
```suggestion
  void on_visibilityCheckBox__toggled(bool checked);
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.h:
```diff
@@ -0,0 +1,30 @@
+  VolumeListWidgetItem(QListWidget* listWidget, QString name, AutoscoperMainWindow* main_window, std::vector< xromm::gpu::RayCaster*>* renderers);
+private:
+  std::vector< xromm::gpu::RayCaster*> renderers_;
+  QCheckBox *visibleCheckBox_;
```
```suggestion
  QCheckBox *visibilityCheckBox_;
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,48 @@
+#include <gpu/opencl/RayCaster.hpp>
+#endif
+
+VolumeListWidgetItem::VolumeListWidgetItem(QListWidget* listWidget,QString name, AutoscoperMainWindow* main_window, std::vector< xromm::gpu::RayCaster*>* renderers) : QListWidgetItem(listWidget) {
```
```suggestion

VolumeListWidgetItem::VolumeListWidgetItem(QListWidget* listWidget, const QString& name, std::vector< xromm::gpu::RayCaster*>* renderers) : 
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,48 @@
+
+VolumeListWidgetItem::VolumeListWidgetItem(QListWidget* listWidget,QString name, AutoscoperMainWindow* main_window, std::vector< xromm::gpu::RayCaster*>* renderers) : QListWidgetItem(listWidget) {
+  this->name_ = name;
+  this->main_window_ = main_window;
```
```suggestion
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.h:
```diff
@@ -0,0 +1,30 @@
+  // List Object to display the volume name and a checkbox to toggle visibility
+  Q_OBJECT
+public:
+  VolumeListWidgetItem(QListWidget* listWidget, QString name, AutoscoperMainWindow* main_window, std::vector< xromm::gpu::RayCaster*>* renderers);
```
```suggestion
  VolumeListWidgetItem(QListWidget* listWidget, const QString& name, std::vector< xromm::gpu::RayCaster*>* renderers);
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.cpp:
```diff
@@ -0,0 +1,48 @@
+  pFrame->setMinimumHeight(32);
+  QGridLayout* pLayout = new QGridLayout(pFrame);
+  pLayout->addWidget(new QLabel(name_), 0, 1);
+  visibleCheckBox_ = new QCheckBox();
+  visibleCheckBox_->setChecked(true);
+  pLayout->addWidget(visibleCheckBox_, 0, 0);
```
```suggestion
  visibilityCheckBox_ = new QCheckBox();
  visibilityCheckBox_->setChecked(true);
  pLayout->addWidget(visibilityCheckBox_, 0, 0);
```

jcfr commented on autoscoper/src/ui/VolumeListWidgetItem.h:
```diff
@@ -0,0 +1,30 @@
+  std::vector< xromm::gpu::RayCaster*> renderers_;
+  QCheckBox *visibleCheckBox_;
+  QString name_;
+  AutoscoperMainWindow*  main_window_;
```
```suggestion
  AutoscoperMainWindow*  main_window_;
```

jcfr commented on autoscoper/src/ui/VolumeDockWidget.cpp:
```diff
@@ -42,12 +44,15 @@ void VolumeDockWidget::clearVol(){
+    renderers.push_back(view->drrRenderer(idx));
+  }
 
+  QListWidgetItem* volumeItem = new VolumeListWidgetItem(dock->listWidget, fi.completeBaseName() ,mainwindow, &renderers);
```
```suggestion
  QListWidgetItem* volumeItem = new VolumeListWidgetItem(dock->listWidget, fi.completeBaseName(), &renderers);
```

NicerNewerCar commented:
@jcfr I addressed most of the suggestions from the last review. However, I haven't removed the pointer to the main window just yet as this is required to update the volumes as soon as the checkbox is toggled. Removing this requires another step done by the user to update the rendering (either selecting a different volume or running the optimizations), which I think is a less fluid experience. We can move forward with this, but I figured I'd show the difference.

| With pointer | Without Pointer |
| ----- | ----- | 
| ![pointer](https://user-images.githubusercontent.com/30351234/271680676-48e75a52-f9e9-492d-9c3a-08008bd877c3.gif) | ![no-main-win-pointer](https://github.com/BrownBiomechanics/Autoscoper/assets/30351234/653c8a39-1a4b-4b1d-8c1c-bde2234962cf) |

amymmorton commented:
with pointer seems the desired option

On Fri, Oct 20, 2023 at 9:56 AM Anthony Lombardi ***@***.***>
wrote:

> @jcfr <https://github.com/jcfr> I addressed most of the suggestions from
> the last review. However, I haven't removed the pointer to the main window
> just yet as this is required to update the volumes as soon as the checkbox
> is toggled. Removing this requires another step done by the user to update
> the rendering (either selecting a different volume or running the
> optimizations), which I think is a less fluid experience. We can move
> forward with this, but I figured I'd show the difference.
> With pointer Without Pointer
> [image: pointer]
> <https://user-images.githubusercontent.com/30351234/271680676-48e75a52-f9e9-492d-9c3a-08008bd877c3.gif> [image:
> no-main-win-pointer]
> <https://user-images.githubusercontent.com/30351234/276952225-653c8a39-1a4b-4b1d-8c1c-bde2234962cf.gif>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/BrownBiomechanics/Autoscoper/pull/209#issuecomment-1772786763>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APUUPFQ5AYYAQEIQN2INZH3YAJ7HBAVCNFSM6AAAAAA5MZIJ46VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONZSG44DMNZWGM>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>


-- 
-----------
Amy Morton, MSc

Sr. Research Engineer, Bioengineering Lab, Department of Orthopaedic
Research
The Warren Alpert Medical School of Brown University and Rhode Island
Hospital
1 Hoppin Street, CORO West, Ste. 404, Providence, RI 02903

jcfr commented:
> removed the pointer to the main window 

Thanks for the explanation. This was on oversight on my side. We will revisit the split UI/Logic at some other time.
